File: solutions/elixir/yacht/lib/yacht.ex
defmodule Yacht do
  @moduledoc false
  import Enum, only: [filter: 2, sort: 1, sum: 1, to_list: 1]

  @type category ::
          :ones
          | :twos
          | :threes
          | :fours
          | :fives
          | :sixes
          | :full_house
          | :four_of_a_kind
          | :little_straight
          | :big_straight
          | :choice
          | :yacht

  @doc """
  Calculate the score of 5 dice using the given category's scoring method.
  """
  @spec score(category :: category(), dice :: [integer]) :: integer
  def score(:ones, dice), do: dice |> sum_for_value(1)
  def score(:twos, dice), do: dice |> sum_for_value(2)
  def score(:threes, dice), do: dice |> sum_for_value(3)
  def score(:fours, dice), do: dice |> sum_for_value(4)
  def score(:fives, dice), do: dice |> sum_for_value(5)
  def score(:sixes, dice), do: dice |> sum_for_value(6)

  # a straight is worth 30 points if all dice values required are present
  def score(:little_straight, dice), do: dice |> score_for(1..5)
  def score(:big_straight, dice), do: dice |> score_for(2..6)
  def score(:choice, dice), do: sum(dice)
  def score(:yacht, [d, d, d, d, d]), do: 50
  def score(:yacht, _dice), do: 0

  def score(:full_house, dice) do
    case sort(dice) do
      [d, d, d, d, d] -> 0
      [s, s, b, b, b] -> sum(dice)
      [s, s, s, b, b] -> sum(dice)
      _ -> 0
    end
  end

  def score(:four_of_a_kind, dice) do
    case sort(dice) do
      [s, s, s, s, _] -> s * 4
      [_, b, b, b, b] -> b * 4
      _ -> 0
    end
  end

  @doc false
  @spec sum_for_value([integer], integer) :: integer
  defp sum_for_value(dice, num),
    do: dice |> filter(&(&1 == num)) |> sum()

  @doc false
  @spec score_for([integer], Range.t()) :: integer
  defp score_for(dice, nums),
    do: if(sort(dice) == to_list(nums), do: 30, else: 0)
end

File: solutions/javascript/yacht/yacht.js
// @ts-check

/**
 * @typedef {'ones' |
 *           'twos' |
 *           'threes' |
 *           'fours' |
 *           'fives' |
 *           'sixes' |
 *           'full house' |
 *           'four of a kind' |
 *           'little straight' |
 *           'big straight' |
 *           'choice' |
 *           'yacht' }  Category;
 */

/**
 * Score a single throw of dice in the game Yacht.
 *
 * @param {number[]} dice - always five dice
 * @param {Category} category for scoring
 * @return {number} score
 */
export const score = (dice, category) => {
  const checkYacht = () => (new Set(dice).size === 1 ? 50 : 0);

  const checkStraight = (/** @type {number} */ start) => {
    dice.sort((a, b) => a - b);
    for (const die of dice) if (die !== start++) return 0;
    // worth 30 points if all required dice values are present
    return 30;
  };

  const checkFourOfAKind = () => {
    const groupsOfDice = determineGroupsOfDice();
    if (groupsOfDice.length > 2) return 0;
    groupsOfDice.sort((a, b) => b.count - a.count);
    return groupsOfDice[0].count >= 4 ? groupsOfDice[0].key * 4 : 0;
  };

  const checkFullHouse = () => {
    const groupsOfDice = determineGroupsOfDice();
    if (groupsOfDice.length !== 2) return 0;
    groupsOfDice.sort((a, b) => b.count - a.count);
    return groupsOfDice[0].count === 3 ? sumAll() : 0;
  };

  const sumAll = () => dice.reduce((sum, die) => (sum += die), 0);

  const sumForValue = (/** @type {number} */ value) =>
    dice.reduce((sum, die) => (sum += die === value ? die : 0), 0);

  /** @typedef {{key: number, count: number}} DiceGroup */
  const determineGroupsOfDice = () =>
    dice.reduce((/** @type {DiceGroup[]} */ result, die) => {
      const group = result.find((g) => g.key === die);
      if (group) group.count += 1;
      else result.push({ key: die, count: 1 });
      return result;
    }, []);

  switch (category) {
    case "ones":
      return sumForValue(1);
    case "twos":
      return sumForValue(2);
    case "threes":
      return sumForValue(3);
    case "fours":
      return sumForValue(4);
    case "fives":
      return sumForValue(5);
    case "sixes":
      return sumForValue(6);
    case "full house":
      return checkFullHouse();
    case "four of a kind":
      return checkFourOfAKind();
    case "little straight":
      return checkStraight(1);
    case "big straight":
      return checkStraight(2);
    case "choice":
      return sumAll();
    case "yacht":
      return checkYacht();
    default:
      return 0;
  }
};
